crates/chia-consensus/src/multiprocess_validation.rs

```diff
@@ -0,0 +1,238 @@
+fn validate_clvm_and_signature(
+    spend_bundle: &SpendBundle,
+    max_cost: u64,
+    constants: ConsensusConstants,
+    height: u32,
+    syncing: bool,
+    cache: Arc<Mutex<BlsCache>>,
+) -> Result<(OwnedSpendBundleConditions, Duration), ErrorCode> {
+    let start_time = Instant::now();
+    let additional_data = constants.agg_sig_me_additional_data;
+    let program: BlockGenerator = simple_solution_generator(spend_bundle)?;
+    let npcresult = get_name_puzzle_conditions(program, max_cost, true, height, &constants)?;
+    let (pks, msgs) = pkm_pairs(npcresult.clone(), additional_data.as_slice())?;
+    // Verify aggregated signature
+    if !{
+        if syncing {
+            // if we're syncing use the chia_bls::aggregate_verify to avoid using the cache
+            let data: Vec<(&PublicKey, &[u8])> = pks
+                .iter()
+                .zip(msgs.iter().map(|msg| msg.as_slice()))
+                .collect();
+            aggregate_verify(&spend_bundle.aggregated_signature, data)
+        } else {
+            // if we're fully synced then use the cache
+            cache
+                .lock()
+                .unwrap()
+                .aggregate_verify(pks, msgs, &spend_bundle.aggregated_signature)
+        }
+    } {
+        return Err(ErrorCode::InvalidSpendBundle);
+    }
+    Ok((npcresult, start_time.elapsed()))
+}
+
+// #[cfg(feature = "py-bindings")]
+// mod py_funcs {
+//     use super::*;
+//     use pyo3::{
+//         exceptions::PyValueError,
+//         pybacked::PyBackedBytes,
+//         pyfunction,
+//         types::{PyAnyMethods, PyList},
+//         Bound, PyObject, PyResult,
+//     };
+//     use crate::gen::owned_conditions;
+
+//     #[pyfunction]
+//     #[pyo3(name = "pre_validate_spendbundle")]
+//     pub fn py_pre_validate_spendbundle(
+//         new_spend: SpendBundle,
+//         max_cost: u64,
+//         constants: ConsensusConstants,
+//         peak_height: u32,
+//         syncing: bool,
+//         cache: BlsCache
+//     ) -> Result<(SpendBundle, OwnedSpendBundleConditions), ErrorCode> {
+//         let sbc = validate_clvm_and_signature(&new_spend, max_cost, constants, peak_height, syncing, Arc::new(Mutex::new(cache)));  // TODO: use cache properly
+//         match sbc {
+//             Ok(owned_conditions) => {
+//                 Ok((new_spend, owned_conditions.0))
+//             },
+//             Err(e) => {
+//                 Err(e)
+//             }
+//         }
+//     }
+// }
+
+pub fn simple_solution_generator(bundle: &SpendBundle) -> Result<BlockGenerator, ErrorCode> {
+    let mut spends = Vec::<(Coin, Vec<u8>, Vec<u8>)>::new();
+    for cs in &bundle.coin_spends {
+        let puzzle_reveal = cs.puzzle_reveal.to_vec();
+        let solution_reveal = cs.solution.to_vec();
+        spends.push((cs.coin, puzzle_reveal, solution_reveal));
+    }
+    let block_program = solution_generator(spends);
+    match block_program {
```

Comment:
Instead of nesting matches like this it's best to use `let`..`else` and early return errors, so there's less indentation.